File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size();

        int len = 0;
        for (int l = 0, r = 0, z = 0; r < n; r++) {
            z += nums[r] == 0;
            if (z > k) {
                z -= nums[l] == 0;
                l++;
            }

            len = max(len, r - l + 1);
        }

        return len;
    }
};

File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: list[int], k: int) -> int:
        N = len(nums)

        largest = 0
        l, z = 0, 0
        for r in range(N):
            z += 1 if nums[r] == 0 else 0
            if z > k:
                z -= 1 if nums[l] == 0 else 0
                l += 1
            else:
                largest = max(largest, r - l + 1)

        return largest